README.md
# DAIICT PC503 Python Project Spam Detection Toolbox

***In the following, we will design a SPAM email detection toolbox. First we will compose the
emails, then we will detect spam. Do not hardcode values/logic when not specified.***<br>

## Part A
1. Create a text file named *student_name_list.txt* of the names of 10 PC503 students manually, keeping one name in each line. Note: this is the only file that is to be created manually. Files in the subsequent questions must not be created manually.<br>

2. Write a Python code *main_data.py* to do the following tasks. Each task must be written in a
function with a suitable function name. Depending on the task, the function may or may not return anything.<br>

3. Read the file *About-DAIICT.txt* and create four new files named *History.txt, Environment.txt, Recognition.txt and Accreditation.txt* by inserting the content of each section as given in *About-DAIICT.txt*.<br>

4. Create a list named *student_name_list* of the 10 names by reading the text file *student_name_list.txt*.<br>

5. Randomly generate 15 files with file names: *email-1.txt* to *email-20.txt*. So, the name of the
13th file need not be *email-13.txt*, it can also be *email-19.txt* or *email-5.txt*. Content of each file (10 lines) should be as follows:<br>
a) First line should have time, date, month, year, email Id in the following format:<br>
*Received at 17:23 hrs on October 7, 2018 from fname_lname@daiict.ac.in*<br>
All entries must be randomly inserted and email Ids can be created from random elements of
*student_name_list*. Obviously, there will be at least one or more files containing the same email Id since the number of students is 10 (<15).<br>
b) Second line should be blank.<br>
c) Third line to the tenth line must be the lines from the four above mentioned text files
in random order such that (i) there is at least one line from each file and (ii) the lines
taken from *History.txt* can include only one of the four sentences containing *“first
wave”, “second wave”, “third wave”, “fourth wave”*. There is no restriction on other
sentences in this text file.<br>

6. Randomly generate 4 more files using the above mentioned logic with two modifications:<br>
a) File name should be the unused numbers between 1 to 20.<br>
b) Email Ids must be taken from the following list: email_ids =[*name1@gmail.com*,
*A_X_y@yahoo.co.in*, *nm123@rediff.com*, *nam_4_e@160.com*].<br>

7. Randomly generate 1 more file using the above mentioned logic with three modifications:<br>
a) File name should be the unused number between 1 to 20.<br>
b) Email Id is *pc_503@daiict.ac.in*<br>
c) The lines taken from *History.txt* cannot include any one of the four sentences
containing *“first wave”, “second wave”, “third wave”, “fourth wave”*. There is no restriction on other sentences in this text file.<br>

## Part B

8. Write a Python code *main_test.py* to do the following tasks. Each task must be written in a
function with a suitable function name. Depending on the task, the function may or may not
return anything.<br>

9. Print the email Ids whose files do not have any *“wave”* information. In the current exercise, there is only one such entry.<br>

10. The users not having daiict email Id or without any *“wave”* information must be spammed.
Rename the corresponding files as *spam1.txt*, *spam2.txt*, … and update their contents by
adding a line at the beginning: *“This email has been categorized as spam”*.<br>

11. Create *ordered_names.txt* containing the user names (non-spam type) in the descending
order of date-time etc., i.e. most recent first. One name in each line. Keep duplicate names.<br>

12. Create *ordered_names_wave1.txt* containing the user names (non-spam type) belonging to
the *“first wave”* group in the descending order of date-time etc., i.e. most recent first. No
duplicate names and one name in each line. Similarly create files for the *“second wave”*,
*“third wave”*, *“fourth wave”* groups.<br>

13. Plot the number of users in each group and number of spam emails using a bar chart.<br>

Note: Randomize the codes so that every time they are executed, different output is expected.


History.txt
Dhirubhai Ambani Institute of Information and Communication Technology (DA-IICT), Gandhinagar represents Wave-4 of educational innovation in Gujarat.

The first wave was the nationalist wave and led to Gandhian experiments in education including Nai Talim. 
The Gujarat Vidyapith established in 1920 was a hybrid model of a University based on Gandhian principles.

The second wave led to the establishment of a whole range of educational institutions in Gujarat, mainly private colleges.

The third wave, an inspired one, was spearheaded by the Industrialist Kasturbhai Lalbhai and the Scientist Vikram Sarabhai and led to the establishment of a network of national institutes of international renown. 
A whole array of remarkable intellectuals provided the leadership of these Institutes.

It was in the fourth wave when the focus shifted to higher education and private participation. One of the Institutes created during this period is DA-IICT in 2001. 
It is the only advanced research and teaching Institute named after the Late Dhirubhai Ambani, the founder of Reliance Group.


main_test.py
import os
import test_data_util as tdu
import re
from datetime import datetime
from matplotlib import pyplot as plt

baseip="main_data_output/"
baseop="test_data/"

# TODO: 9. Print the email Ids whose files do not have any “wave” information.
#   In the current exercise, there is only one such entry.


# TODO: 10. The users not having daiict email Id or without any “wave” information must be spammed.
#   Rename the corresponding files as spam1.txt, spam2.txt, … and update their contents by
#   adding a line at the beginning: “This email has been categorized as spam”.


# TODO: 11. Create ordered_names.txt containing the user names (non-spam type) in the descending
#   order of date-time etc., i.e. most recent first. One name in each line. Keep duplicate names.
def Task11():
    fileName="ordered_names.txt"
    global baseip
    listMail=tdu.find_file()
    dateName=tdu.extractDateName(listMail)
    sorted(dateName,key=lambda date: datetime.strptime(date,"%H:%M hrs on %B %d, %Y"))

    with open(os.path.join(baseop,fileName),"w") as fp:
        for i in dateName:
            fp.write(dateName[i]+"\n")

# TODO: 12. Create ordered_names_wave1.txt containing the user names (non-spam type) belonging to
#   the “first wave” group in the descending order of date-time etc., i.e. most recent first. No
#   duplicate names and one name in each line. Similarly create files for the “second wave”,
#   “third wave”, “fourth wave” groups.
def Task12():
    fileName="ordered_names_wave1.txt"
    listMail=tdu.find_file()
    #List to store all the filename who is containing relative information
    listWave1=[]
    listWave2=[]
    listWave3=[]
    listWave4=[]

    for file in listMail:
        wave=tdu.findWave(file) 
        if(wave=='1'):
            listWave1.append(file)
        elif(wave=='2'):
            listWave2.append(file)
        elif(wave=='3'):
            listWave3.append(file)
        elif(wave=='4'):
            listWave4.append(file)
    
    with open(os.path.join(baseop,"ordererd_names_wave1.txt"),"w") as fp1:
        orderedWave1=tdu.extractDateName(listWave1)
        for name in orderedWave1.values():
            fp1.writelines(name+"\n")
    
    with open(os.path.join(baseop,"ordererd_names_wave2.txt"),"w") as fp1:
        orderedWave2=tdu.extractDateName(listWave2)
        for name in orderedWave2.values():
            fp1.writelines(name+"\n")

    with open(os.path.join(baseop,"ordererd_names_wave3.txt"),"w") as fp1:
        orderedWave3=tdu.extractDateName(listWave3)
        for name in orderedWave3.values():
            fp1.writelines(name+"\n")

    with open(os.path.join(baseop,"ordererd_names_wave4.txt"),"w") as fp1:
        orderedWave4=tdu.extractDateName(listWave4)
        for name in orderedWave4.values():
            fp1.writelines(name+"\n")

# TODO: 13. Plot the number of users in each group and number of spam emails using a bar chart.
def Task13():
    data=dict()

    for i in 
    fp1=open(os.path.join(baseop,"ordererd_names_wave1.txt"),"r")
    tmpList=fp1.readlines()
    data["Wave1"]=len(tmpList)

    fp1=open(os.path.join(baseop,"ordererd_names_wave2.txt"),"r")
    tmpList=fp1.readlines()
    data["Wave2"]=len(tmpList)

    fp1=open(os.path.join(baseop,"ordererd_names_wave3.txt"),"r")
    tmpList=fp1.readlines()
    data["Wave3"]=len(tmpList)

    fp1=open(os.path.join(baseop,"ordererd_names_wave4.txt"),"r")
    tmpList=fp1.readlines()
    data["Wave4"]=len(tmpList)

    #for spam file
    tmpList=os.listdir("main_data_output/")
    r=re.compile("spam.*")
    tmpList=list(filter(r.match,tmpList))
    # data["Spam"]=len(list)

    fig=plt.figure(figsize=(7,5)).canvas.set_window_title("Output")
    graph=plt.bar(data.keys(),data.values(),color="maroon",width=0.3)
    graph[4].set_color("blue")
    plt.xlabel("Wave category")
    plt.ylabel("Frequency")
    plt.title("Frequency of wave statements")
    plt.show()

if __name__ == "__main__":
    Task13()

test_data_util.py
import os
import re

baseip="main_data_output/"
#Find number of files in folder with that name email-*.txt and generate list of that
def find_file():
    listAll=os.listdir("main_data_output/")
    r=re.compile("email.*")
    listMail=list(filter(r.match,listAll))

    return listMail

#i/p: List of files o/p:Will get dictonary, key:Time Value: Username
def extractDateName(listMail):
    dateName=dict()
    for file in listMail:
        path=os.path.join(baseip,file)
        with open(path,"r") as Efile:
            str=Efile.readline()
            data=re.findall("[0-9].*",str) 
            for i in range(len(data)):
                x=data[i].split(" from ")
                x[1]=x[1].replace("@daiict.ac.in","")
                dateName[x[0]]=x[1]
    return dateName

def findWave(fname):
    with open(os.path.join(baseip,fname),"r") as fp:
        data=fp.readlines()
        r=re.compile(".* wave")
        listData=list(filter(r.match,data))
        waveType=0
        if(len(listData)!=0):
            x=listData[0].split("wave")
            tmp=x[0].strip()
            tmplist=tmp.split(" ")

            if(tmplist[-1] == "first"):
                waveType="1"
            elif(tmplist[-1] == "second"):
                waveType="2"
            elif(tmplist[-1] == "third"):
                waveType="3"
            elif(tmplist[-1] == "fourth"):
                waveType="4"
    return waveType

main_data_util.py
import random
import datetime


# Random date-time generator helper function
def dateTimeGenerator():
    datetime_list = []
    for i in range(20):
        month = random.randint(1, 12)
        day = random.randint(1, 28)
        hours = random.randint(0, 23)
        minutes = random.randint(0, 59)
        datetime_list.append(datetime.datetime(2020, month, day, hours, minutes).strftime("%H:%M hrs on %B %d, %Y"))

    return datetime_list


# Extract lines from Accreditation.txt file
def __extractAccreditationLines():
    accreditation = open("main_data_output/Accreditation.txt", "r")
    acc_lines = []
    for line in accreditation.readlines():
        if line == "\n":
            continue
        acc_lines.append(line.strip())
    accreditation.close()
    return acc_lines


# Extract lines from Recognition.txt file
def __extractRecognitionLines():
    recognition = open("main_data_output/Recognition.txt", "r")
    recog_lines = []
    for line in recognition.readlines():
        if line == "\n":
            continue
        recog_lines.append(line.strip())
    recognition.close()
    return recog_lines


# Extract lines from Environment.txt file
def __extractEnvironmentLines():
    environment = open("main_data_output/Environment.txt", "r")
    env_lines = []
    for line in environment.readlines():
        if line == "\n":
            continue
        env_lines.append(line.strip())
    environment.close()
    return env_lines


# Extract and return list of wave lines from the History.txt file
def __extractWaveNonWaveLines():
    history = open("main_data_output/History.txt", "r")
    data = history.readlines()
    extracted_waves = []
    for i in range(len(data)):
        if data[i].startswith('Dhirubhai') or data[i].startswith('\n'):
            continue
        extracted_waves.append(data[i].strip())

    waves = []
    i = 0
    while i < len(extracted_waves) - 1:
        if extracted_waves[i].startswith("The second wave"):
            waves.append(extracted_waves[i])
            i += 1
            continue
        waves.append(extracted_waves[i])
        i += 2

    not_waves = [data[0].strip()]  # Initializing with the first line of History.txt
    for line in extracted_waves:
        if line not in waves:
            not_waves.append(line)

    split_fourth_wave = waves[len(waves) - 1].split(". ")  # Since there are two sentences in one line
    waves.pop()
    waves.append(split_fourth_wave[0] + ".")
    not_waves.append(split_fourth_wave[1])
    history.close()
    return waves, not_waves


# Extract lines from Accreditation, Environment, Recognition text files
# Make a large list of lines from all files except wave lines and another list of wave lines
# and return a random choice of 8 lines - 1 from wave lines and 7 from the other lines
def random8LinesGenerator(has_wave=True):
    wave_lines, non_wave_lines = __extractWaveNonWaveLines()
    env_lines = __extractEnvironmentLines()
    acc_lines = __extractAccreditationLines()
    recog_lines = __extractRecognitionLines()

    eight_lines = []
    # Selecting random 8 lines from the corresponding file lists
    # and ensuring there is at least one line from each file list
    if has_wave:
        combined_lines = acc_lines + env_lines + recog_lines

        eight_lines.append(random.choice(wave_lines))

        eight_lines.append(random.choice(env_lines))
        eight_lines.append(random.choice(acc_lines))
        eight_lines.append(random.choice(recog_lines))

        eight_lines.extend(random.sample(combined_lines, 4))
    else:
        combined_lines = non_wave_lines + acc_lines + env_lines + recog_lines

        eight_lines.append(random.choice(non_wave_lines))
        eight_lines.append(random.choice(env_lines))
        eight_lines.append(random.choice(acc_lines))
        eight_lines.append(random.choice(recog_lines))

        eight_lines.extend(random.sample(combined_lines, 4))

    random.shuffle(eight_lines)
    # Some post-processing
    for i in range(len(eight_lines) - 1):
        eight_lines[i] = eight_lines[i] + "\n"

    assert len(eight_lines) == 8

    # for line in eight_lines:
    #     print(line)
    return eight_lines


main_data.py
import random
import main_data_util as mdu


# Creating 4 files from About-DAIICT.txt
def generateHERAFiles():
    with open("About-DAIICT.txt", "r") as aboutDaiict:
        data = aboutDaiict.readlines()

        # Create files
        history = open("main_data_output/History.txt", "w")
        environment = open("main_data_output/Environment.txt", "w")
        recognition = open("main_data_output/Recognition.txt", "w")
        accreditation = open("main_data_output/Accreditation.txt", "w")

        # Update appropriate text to files by iterating over data
        hisIdx, envIdx = data.index("History\n"), data.index("Environment\n")
        recogIdx, accredIdx = data.index("Recognition\n"), data.index("Accreditation\n")
        for i in range(len(data)):
            if hisIdx + 1 < i < envIdx:
                history.write(data[i])
            if envIdx + 1 < i < recogIdx:
                environment.write(data[i])
            if recogIdx + 1 < i < accredIdx:
                recognition.write(data[i])
            if i > accredIdx + 1:
                accreditation.write(data[i])

        history.close()
        environment.close()
        recognition.close()
        accreditation.close()
        # print('data.index("History\\n"): ', data.index("History\n"))
        # print('data.index("Environment\\n"): ', data.index("Environment\n"))
        # print('data.index("Recognition\\n"): ', data.index("Recognition\n"))
        # print('data.index("Accreditation\\n"): ', data.index("Accreditation\n"))


# Create and return list of names of students
def createStudentNameList():
    with open("student_name_list.txt", "r") as stuNameFile:
        student_name_list = [name.rstrip() for name in stuNameFile.readlines()]
    return student_name_list


# Generating email text files as specified in todos 5, 6, 7
def generateEmailFiles(student_name_list):
    # Generate email text file numbers and shuffle them for randomness
    emailNums = [i for i in range(1, 21)]
    random.shuffle(emailNums)
    print(emailNums)

    # Creating datetime list for randomising date-time in file
    datetime_list = mdu.dateTimeGenerator()

    # Task 5
    # Creating a new list by extending the original list by randomly chosen 5 names from the same list
    # and shuffling it for more randomness
    student_list_extended = student_name_list + random.sample(student_name_list, 5)
    random.shuffle(student_list_extended)
    print(student_list_extended)

    # Looping and generating files
    for i in range(len(emailNums) - 5):
        emailTxt = open(f"main_data_output/email-{emailNums[i]}.txt", "w")
        # Create email id
        split_student_name = student_list_extended[i].split(" ")
        assert len(split_student_name) == 2
        email_id = f"{split_student_name[0].lower()}_{split_student_name[1].lower()}@daiict.ac.in"

        # Write to newly generated file
        emailTxt.write(f"Received at {datetime_list[i]} from {email_id}\n")  # 5.(a)
        emailTxt.write("\n")  # 5.(b)
        for line in mdu.random8LinesGenerator():  # 5.(c)
            emailTxt.write(line)

        emailTxt.close()

    # Task 6
    # Generating spam email id text files with the same content as above generated files
    spam_emails = ["name1@gmail.com", "A_X_y@yahoo.co.in", "nm123@rediff.com", "nam_4_e@160.com"]
    random.shuffle(spam_emails)

    # Generating files with the unused 4 numbers
    offset = 15  # Offset for indexing unused numbers in emailNums list
    for i in range(len(spam_emails)):
        emailTxt = open(f"main_data_output/email-{emailNums[i + offset]}.txt", "w")

        # Write to newly generated file
        emailTxt.write(f"Received at {datetime_list[i+offset]} from {spam_emails[i]}\n")
        emailTxt.write("\n")
        for line in mdu.random8LinesGenerator():
            emailTxt.write(line)

        emailTxt.close()

    # Task 7
    # Generating one last email text file using the last unused number and a given email id
    # and it will not have any wave lines
    last_email_num, last_date_time = emailNums[-1], datetime_list[-1]
    last_email_id = "pc_503@daiict.ac.in"

    emailTxt = open(f"main_data_output/email-{last_email_num}.txt", "w")
    emailTxt.write(f"Received at {last_date_time} from {last_email_id}\n")
    emailTxt.write("\n")
    for line in mdu.random8LinesGenerator(has_wave=False):  # Setting has_wave to False to not include any wave line
        emailTxt.write(line)

    emailTxt.close()


if __name__ == "__main__":
    generateHERAFiles()
    generateEmailFiles(createStudentNameList())
